Content.Server/Changeling/ChangelingSystem.cs

```diff
@@ -0,0 +1,624 @@
+            return false;
+        }
+
+        if (!lingAction.UseInLesserForm && comp.IsInLesserForm)
+        {
+            _popup.PopupEntity(Loc.GetString("changeling-action-fail-lesserform"), uid, uid);
+            return false;
+        }
+
+        if (comp.Chemicals < lingAction.ChemicalCost)
+        {
+            _popup.PopupEntity(Loc.GetString("changeling-chemicals-deficit"), uid, uid);
+            return false;
+        }
+
+        if (lingAction.RequireAbsorbed > comp.TotalAbsorbedEntities)
+        {
+            var delta = lingAction.RequireAbsorbed - comp.TotalAbsorbedEntities;
+            _popup.PopupEntity(Loc.GetString("changeling-action-fail-absorbed", ("number", delta)), uid, uid);
+            return false;
+        }
+
+        UpdateChemicals(uid, comp, -lingAction.ChemicalCost);
+        UpdateBiomass(uid, comp, -lingAction.BiomassCost);
+
+        action.Handled = true;
+
+        return true;
+    }
+    public bool TrySting(EntityUid uid, ChangelingComponent comp, EntityTargetActionEvent action, bool overrideMessage = false)
+    {
+        if (!TryUseAbility(uid, comp, action))
+            return false;
+
+        var target = action.Target;
+
+        // can't get his dna if he doesn't have it!
+        if (!HasComp<AbsorbableComponent>(target) || HasComp<AbsorbedComponent>(target))
+        {
+            _popup.PopupEntity(Loc.GetString("changeling-sting-extract-fail"), uid, uid);
+            return false;
+        }
+
+        if (HasComp<ChangelingComponent>(target))
+        {
+            _popup.PopupEntity(Loc.GetString("changeling-sting-fail-self", ("target", Identity.Entity(target, EntityManager))), uid, uid);
+            _popup.PopupEntity(Loc.GetString("changeling-sting-fail-ling"), target, target);
+            return false;
+        }
+        if (!overrideMessage)
+            _popup.PopupEntity(Loc.GetString("changeling-sting", ("target", Identity.Entity(target, EntityManager))), uid, uid);
+        return true;
+    }
+    public bool TryInjectReagents(EntityUid uid, List<(string, FixedPoint2)> reagents)
+    {
+        var solution = new Solution();
+        foreach (var reagent in reagents)
+            solution.AddReagent(reagent.Item1, reagent.Item2);
+
+        if (!_solution.TryGetInjectableSolution(uid, out var targetSolution, out var _))
+            return false;
+
+        if (!_solution.TryAddSolution(targetSolution.Value, solution))
+            return false;
+
+        return true;
+    }
+    public bool TryReagentSting(EntityUid uid, ChangelingComponent comp, EntityTargetActionEvent action, List<(string, FixedPoint2)> reagents)
+    {
+        var target = action.Target;
+        if (!TrySting(uid, comp, action))
+            return false;
+
+        if (!TryInjectReagents(target, reagents))
+            return false;
+
+        return true;
+    }
+    public bool TryToggleItem(EntityUid uid, EntProtoId proto, ChangelingComponent comp, string? clothingSlot = null)
```

Comment:
I will be very glad if you make a separate system for this that can be used not only for changeling.
At the very least, I need it for the cult of flesh and also for the IPC charging module.